File: HasherBase.pas
unit HasherBase;
//Author: domasz
//Version: 0.1 (2022-11-17)
//Licence: MIT

interface

uses SysUtils, Classes, FGL;

type

  { THasherBase }

  THasherBase = class
  public
    Check: String; //hash of '123456789'
    constructor Create; virtual;
    procedure Update(Msg: PByte; Length: Integer); virtual; abstract;
    function Final: String; virtual; abstract;
  end;

  { THasherList }

  THasherClass = class of THasherBase;
  THasherMap = specialize TFPGmap<string, THasherClass>;

  THasherList = class
  private
    FList: THasherMap;
  public
    constructor Create;
    destructor Destroy; override;
    function Count: Integer;
    function GetName(Index: Integer): String;
    function FindClass(Name: String; out AClass: THasherClass): Boolean;
    procedure RegisterHasher(Name: String; AClass: THasherClass);
  end;

  var HasherList: THasherList;

implementation

{ THasherBase }

constructor THasherBase.Create;
begin
  inherited Create;
end;

{ THasherList }

constructor THasherList.Create;
begin
  inherited Create;
  FList := THasherMap.Create;
end;

destructor THasherList.Destroy;
begin
  FList.Free;
  inherited;
end;

function THasherList.Count: Integer;
begin
  Result := FList.Count;
end;

function THasherList.GetName(Index: Integer): String;
begin
  if (Index > FList.Count-1) or (Index < 0) then Result := ''
  else                                           Result := FList.Keys[Index];
end;

function THasherList.FindClass(Name: String; out AClass: THasherClass): Boolean;
var Index: Integer;
begin
  Name := LowerCase(Name);
  Index := FList.IndexOf(Name);

  if Index < 0 then begin
    Result := False;
    Exit;
  end;

  AClass := FList.Data[Index];
  Result := True;
end;

procedure THasherList.RegisterHasher(Name: String; AClass: THasherClass);
begin
  FList.Add(LowerCase(Name), AClass);
end;

initialization
  HasherList := THasherList.Create;

finalization
  HasherList.Free;

end.

File: PHP/HasherBase.php
<?php
//Author: domasz
//Version: 0.1 (2022-11-19)
//Licence: MIT

class HasherBase
{
	public $Check = ''; //hash of '123456789'
}

class HasherList
{
	protected $List = array();
	
	public function Count()
	{
		return count($this->List);
	}

	public function GetName($Index)
	{
		$keys = array_keys($this->List);

		if (($Index > count($keys)-1) or ($Index < 0)) return '';;

		return $keys[$Index];  
	}

	public function FindClass($Name, &$AClass)
	{
		$Name = strtolower($Name);

		$Index = isset($this->List[$Name]);

		if (!$Index)
		{
			return false;
		}

		$AClass = $this->List[$Name];
		return true;
	}

	public function RegisterHasher($Name, $AClass)
	{
		$Name = strtolower($Name);
		$this->List[$Name] = $AClass;
	}
}

$HasherList = new HasherList;